### PR TITLE
(Closes #2913) fix for frontend scoping bug

### LINKED ISSUE #2913: [PSyIR] fparser2 frontend does not respect scoping when resolving the targets of Calls.
Given the code:
```fortran
    module this_mod
      implicit none
    contains
      subroutine do_it()
        use my_mod
        integer :: old_my_sub
        real, dimension(10) :: a
        call my_sub(a)
      end subroutine do_it
      subroutine my_sub()
        use my_mod, only: my_sub
        real, dimension(10) :: b
        b = 1.0
      end subroutine my_sub
    end module this_mod
```
we can't be sure whether the `call` to `my_sub` will resolve to the routine of that name in the same module or whether it is satisfied by the `use my_mod` wildcard import in scope at the callsite. Unfortunately, we currently default to having it linked to the routine inside the module. This is wrong - it should have an UnresolvedInterface since, without resolving the symbols provided by `my_mod`, we don't know where it's from.

### PR DESCRIPTION
**Pull request by arporter**

### CODE REVIEW

**Comment by arporter** [comment]
This is ready for review now (assuming the integration tests are happy). It's a small but important change to the fparser2 frontend so that if we see a wildcard import, we now allow for the fact that it may be responsible for bringing an unresolved Symbol into scope. Previously, we carried on searching in outer scopes. I also had to extend SymbolTable.wildcard_imports so that it has a `scope_limit` flag as currently it automatically searches through all ancestor scopes.

One for @sergisiso, @LonelyCat124 or @hiker I think.

**Review by sergisiso — CHANGES_REQUESTED** [review]
@arporter I agree that we need the proposed solution, but since could leave symbols split (the outer scope and the inner having different symbols that actually should be the same), and break some functionality (e.g. rename outer scope symbol), could we leverage additional information (see inline) for doing better when we can guarantee that they are in fact the same symbol?

File: src/psyclone/tests/domain/common/transformations/raise_psyir_2_alg_trans_test.py [hunk 1]
```diff
@@ -329,17 +329,20 @@ def test_arg_error(fortran_reader, arg):
...
-        f"subroutine alg()\n"
+        f"subroutine alg1()\n"
         f"  use kern_mod\n"
         f"  use field_mod, only : r2d_field\n"
         f"  type(r2d_field) :: field\n"
+        # Persuade the PSyIR that `alg` is a RoutineSymbol.
+        f"  call alg(field)\n"
         f"  call invoke({arg})\n"
-        f"end subroutine alg\n")
+        f"end subroutine alg1\n")
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/domain/common/transformations/raise_psyir_2_alg_trans_test.py:339]: Possibly "testmod" could also not be shadowed?

File: src/psyclone/psyir/frontend/fparser2.py [hunk 2]
```diff
@@ -295,19 +296,31 @@ def _find_or_create_unresolved_symbol(location, name, scope_limit=None,
...
+        if sym:
+            if "symbol_type" in kargs:
+                expected_type = kargs.pop("symbol_type")
+                if not isinstance(sym, expected_type):
+                    # The caller specified a sub-class so we need to
+                    # specialise the existing symbol.
+                    sym.specialise(expected_type, **kargs)
+            return sym
+
+        if table.wildcard_imports(scope_limit=table.node):
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/frontend/fparser2.py:314]: *I suppose we don't do transitive imports, so it's not true that we bring all symbols into the scope

File: src/psyclone/psyir/frontend/fparser2.py [hunk 3]
```diff
@@ -295,19 +296,31 @@ def _find_or_create_unresolved_symbol(location, name, scope_limit=None,
...
-                sym.specialise(expected_type, **kargs)
-        return sym
-    except KeyError:
-        pass
-
-    # find the closest ancestor symbol table attached to a Routine or
+    table = location.scope.symbol_table
+    while table:
+        # By default, `lookup` looks in all ancestor scopes. However, we need
+        # to allow for wildcard imports as we work our way up.
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/frontend/fparser2.py:302]: Done.

File: src/psyclone/psyir/symbols/symbol_table.py [hunk 4]
```diff
@@ -1864,7 +1863,7 @@ def rename_symbol(self, symbol, name, dry_run=False):
         # Re-insert modified symbol
         self.add(symbol)
 
-    def wildcard_imports(self):
+    def wildcard_imports(self, scope_limit=None):
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/symbols/symbol_table.py:1866]: Oops. I tried to do typehinting but couldn't import `Node` (without a circular dep) so gave up.

File: src/psyclone/tests/domain/common/transformations/raise_psyir_2_alg_trans_test.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/domain/common/transformations/raise_psyir_2_alg_trans_test.py:339]: So, if we're asked to look for a Symbol and it turns out we are inside a Routine of the same name then the reference must be to that Routine *unless* it is a Fortran function that contains a DataSymbol of the same name as the result. In this case the standard says:
image
so we must return the DataSymbol in this case.

[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/domain/common/transformations/raise_psyir_2_alg_trans_test.py:339]: I've added some code to do this. It's probably a fairly rare edge case for us though.

File: src/psyclone/psyir/frontend/fparser2.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/frontend/fparser2.py:314]: Yes, I've created #2915 and added a TODO here.

**Comment by arporter** [comment]
All tests are now passing but NEMO4 is not:
```
$ python -m pdb /home/Projects/PSyclone/bin/psyclone -s /home/Projects/PSyclone/examples/nemo/scripts/acc_kernels_trans.py -I /home/Projects/NEMO/NEMO/tests/BENCH_ACC_KERNELS_NVHPC/BLD_SCT_PSYCLONE/ppsrc/nemo /home/Projects/NEMO/NEMO/tests/BENCH_ACC_KERNELS_NVHPC/BLD_SCT_PSYCLONE/ppsrc/nemo/icethd_sal.f90
```

**Comment by arporter** [comment]
Thanks for the review Sergi. I've tightened things up a little bit but, unless I've missed something, there's not much I could do. Assuming the CI is now happy this is ready for another look.

File: src/psyclone/tests/domain/common/transformations/raise_psyir_2_alg_trans_test.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/domain/common/transformations/raise_psyir_2_alg_trans_test.py:339]: > It's probably a fairly rare edge case for us though.

Doesn't it happen to all algorithm examples you had to change, now this can be reverted to:
```
    '''
    code = (
        f"subroutine alg()\n"
        f"  use kern_mod\n"
        f"  use field_mod, only : r2d_field\n"
        f"  type(r2d_field) :: field\n"
        f"  call invoke({arg})\n"
        f"end subroutine alg\n")

    psyir = fortran_reader.psyir_from_source(code)
    call = psyir.walk(Call)[0]
```
Can you revert all tests you had to modify for this?

**Review by sergisiso — CHANGES_REQUESTED** [review]
I didn't do a full review but one of my comments was only partially addressed, can you have a look at reverting the modified tests?

File: src/psyclone/tests/domain/common/transformations/raise_psyir_2_alg_trans_test.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/domain/common/transformations/raise_psyir_2_alg_trans_test.py:339]: Oh yes, I see what you mean, sorry. Reverted.

[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/domain/common/transformations/raise_psyir_2_alg_trans_test.py]: I haven't reverted *all* of the changes to the tests because I think some of them were genuinely wrong before and I've corrected them. However, I have reverted all those where I had to poke it to make the PSyIR think it had a routine symbol.

**Comment by arporter** [comment]
Ready again now.

**Review by sergisiso — APPROVED** [review]
All comments have been addressed, all test are green (I will launch integration once more), the PR is a bugfix so it doesn't need extra documentation. This is approved for merging @arporter
